libs/BLEKit/include/BLEServiceReceiveFile.h

```diff
@@ -0,0 +1,60 @@
+// Leka - LekaOS
+// Copyright 2022 APF France handicap
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <filesystem>
+
+#include "internal/BLEService.h"
+#include "internal/ServicesCharacteristics.h"
+
+namespace leka {
+
+class BLEServiceReceiveFile : public interface::BLEService
+{
+  public:
+	BLEServiceReceiveFile() : interface::BLEService(service::receive_file::uuid, _characteristic_table) {};
+
+	[[nodiscard]] auto getFilePath() const -> std::array<char, 260> { return file_path; }
+
+	void onDataReceived(const data_received_handle_t &params) final
+	{
+		if (params.handle == file_path_characteristic.getValueHandle()) {
+			if (params.offset == 0) {
+				file_path.fill('\0');
+			}
+			std::copy(params.data, params.data + params.len, file_path.begin() + params.offset);
+		}
+		if (params.handle == file_reception_stream_characteristic.getValueHandle()) {
+			if (params.offset == 0) {
+				file_reception_stream.fill('\0');
+			}
+			std::copy(params.data, params.data + params.len, file_reception_stream.begin() + params.offset);
+			if (_on_stream_callback) {
+				_on_stream_callback(std::span {file_reception_stream.data(), params.len});
+			}
```

Comment:
Well, it works in real life

Changing the type of buffer will not change the shortcoming of data receive rate that depends only from the source. If the source set the data rate to 1MB/s, either the classical array and the circular buffer will struggle the same way.

A efficient way to handle this is to send back a signal to the source to inform that the received data has been processed.

I applied KISS for a first attempt since there was no evidence that will work and since there isn't feature in iPad to proceed it (for now).

I will change as you suggested with a circular buffer, but it will take time as it has to be tested to check that it works in real life.